src/Npgsql/Internal/TypeMapping/TypeMapper.cs

```diff
@@ -69,9 +71,26 @@ internal TypeMapper(NpgsqlConnector connector, INpgsqlNameTranslator defaultName
         _databaseInfo = databaseInfo;
 
         var resolvers = new TypeHandlerResolver[resolverFactories.Count];
+        var mapperResolvers = new List<TypeMapperResolver>(resolverFactories.Count);
         for (var i = 0; i < resolverFactories.Count; i++)
+        {
             resolvers[i] = resolverFactories[i].Create(this, Connector);
-        _resolvers = resolvers;
+            var mapperResolver = resolverFactories[i].CreateMapperResolver();
+            if (mapperResolver is not null)
+                mapperResolvers.Add(mapperResolver);
+        }
+
+        // Add global mapper resolvers in backwards because they're inserted in the beginning
+        for (var i = resolverFactories.Count - 1; i >= 0; i--)
+        {
+            
+            var globalMapperResolver = resolverFactories[i].CreateGlobalMapperResolver();
+            if (globalMapperResolver is not null)
+                GlobalTypeMapper.Instance.TryAddMapperResolver(globalMapperResolver);
```

Comment:
Doesn't this mean that if I add some plugin to my specific data source, we change the global type mapper? If so, this is new behavior that I'm not sure we want - the global type mapper config is supposed to flow to the data source, but not vice versa, no? At least up to now, the idea was that you only see mappings on the global type mapper for plugins which you add to it.

See general review comment on where we want to go with NpgsqlParameter type inference.

Reply:
> Doesn't this mean that if I add some plugin to my specific data source, we change the global type mapper? If so, this is new behavior that I'm not sure we want - the global type mapper config is supposed to flow to the data source, but not vice versa, no? At least up to now, the idea was that you only see mappings on the global type mapper for plugins which you add to it.

Yesn't :trollface:
The main problem are ranges (again). In 7.0 we were able to map on them via `GlobalTypeMapper` automatically (as it was done in `BuiltInTypeResolver` and `TypeMapper`). Currently, that's not going to work as (almost) all of that logic was moved to a separate resolver. And we don't really want to add mapping resolver to `GlobalTypeMapper` because it references `NpgsqlRange`, which brings a lot of code + trimming warning.

Reply:
I think that if we need to, it's not the end of the world to not have ranges in the global type mapper. We already don't take into account data source mappings, so it's not like it's a full-featured thing.

Reply:
Good point, though at least in some other cases having a type mapping resolver rooted globally maybe isn't so bad, since the types it references aren't problematic...

So you think it's better to just have the data source mappings impact the global type mapper, with the assumption that there's only one data source in the typical app anyway?

Reply:
> So you think it's better to just have the data source mappings impact the global type mapper, with the assumption that there's only one data source in the typical app anyway?

That pretty much depends on whether we're going to support global type mapper. I personally don't really like this design (having to differentiate between a mapper for local and a mapper for global type mappers, adding a specific mapper to global automatically), but if we're going to support global type mapper (at least for the foreseeable future), we might as well merge this and refactor later (removing the code shouldn't be that hard).
And the my assumption was that aren't going to use multiple data sources with different mappings (e.g. one has ranges, one doesn't, that's probably only going to be true for custom types, which we don't map anyway).